src/MongoDB.Driver.Core/Core/Compression/Zstd/ZstandardNativeWrapper.cs

```diff
@@ -0,0 +1,292 @@
+                    _recommendedZstreamOutputSize = Zstandard64NativeMethods.ZSTD_CStreamOutSize();
+                    _zstreamPointer = Zstandard64NativeMethods.ZSTD_createCStream(); // create resource
+                    break;
+                case CompressionMode.Decompress:
+                    // calculate recommended size for input buffer
+                    _recommendedZstreamInputSize = Zstandard64NativeMethods.ZSTD_DStreamInSize();
+                    // calculate recommended size for output buffer. Guarantee to successfully flush at least one complete block in all circumstances
+                    _recommendedZstreamOutputSize = Zstandard64NativeMethods.ZSTD_DStreamOutSize();
+                    _zstreamPointer = Zstandard64NativeMethods.ZSTD_createDStream(); // create resource
+                    break;
+            }
+        }
+
+        public int RecommendedInputSize => (int)_recommendedZstreamInputSize;
+        public int RecommendedOutputSize => (int)_recommendedZstreamOutputSize;
+
+        // public methods
+        public void Compress(
+            OperationContext operationContext,
+            int compressedSize,
+            int uncompressedSize,
+            out int compressedBufferPosition,
+            out int uncompressedBufferPosition)
+        {
+            Ensure.IsNotNull(operationContext, nameof(operationContext));
+
+            InitializeIfNotAlreadyInitialized();
+
+            // compressed data
+            ConfigureNativeBuffer(
+                _outputNativeBuffer,
+                operationContext.CompressedPinnedBuffer, // operation result
+                (uint)compressedSize);
+
+            // uncompressed data
+            ConfigureNativeBuffer(
+                _inputNativeBuffer,
+                operationContext.UncompressedPinnedBuffer,
+                (uint)uncompressedSize);
+
+            // compress _inputNativeBuffer to _outputNativeBuffer
+            Zstandard64NativeMethods.ZSTD_compressStream(
+                _zstreamPointer,
+                outputBuffer: _outputNativeBuffer,
+                inputBuffer: _inputNativeBuffer);
+
+            compressedBufferPosition = (int)_outputNativeBuffer.Position;
+
+            // calculate progress in _inputNativeBuffer
+            uncompressedBufferPosition = (int)_inputNativeBuffer.Position;
+            operationContext.UncompressedPinnedBuffer.Offset += uncompressedBufferPosition;
+            // CompressedPinnedBuffer.Offset is always 0
+        }
+
+        public void Decompress(
+            OperationContext operationContext,
+            int compressedOffset,
+            int compressedSize,
+            int uncompressedSize,
+            out int compressedBufferPosition,
+            out int uncompressedBufferPosition)
+        {
+            Ensure.IsNotNull(operationContext, nameof(operationContext));
```

Comment:
Check that calling `Decompress` is valid?

That is, that the `CompressionMode` is` Decompress`.